federation/scim/src/main/java/org/keycloak/federation/scim/core/service/UserScimService.java

```diff
@@ -0,0 +1,130 @@
+                .map(this::getId);
+        Optional<KeycloakId> matchedByEmail = resource.getEmails().stream().findFirst().flatMap(MultiComplexNode::getValue)
+                .map(getKeycloakDao()::getUserByEmail).map(this::getId);
+        if (matchedByUsername.isPresent() && matchedByEmail.isPresent() && !matchedByUsername.equals(matchedByEmail)) {
+            String inconstencyErrorMessage = "Found 2 possible users for remote user " + matchedByUsername.get() + " - "
+                    + matchedByEmail.get();
+            LOGGER.warn(inconstencyErrorMessage);
+            throw new InconsistentScimMappingException(inconstencyErrorMessage);
+        }
+        if (matchedByUsername.isPresent()) {
+            return matchedByUsername;
+        }
+        return matchedByEmail;
+    }
+
+    @Override
+    protected KeycloakId createEntity(User resource) throws UnexpectedScimDataException {
+        String username = resource.getUserName().filter(StringUtils::isNotBlank)
+                .orElseThrow(() -> new UnexpectedScimDataException(
+                        "Remote Scim user has empty username, can't create. Resource id = %s".formatted(resource.getId())));
+        UserModel user = getKeycloakDao().addUser(username);
+        resource.getEmails().stream().findFirst().flatMap(MultiComplexNode::getValue).ifPresent(user::setEmail);
+        boolean userEnabled = resource.isActive().orElse(false);
+        user.setEnabled(userEnabled);
+        return new KeycloakId(user.getId());
+    }
+
+    @Override
+    protected boolean isMarkedToIgnore(UserModel userModel) {
+        return BooleanUtils.TRUE.equals(userModel.getFirstAttribute("scim-skip"));
+    }
+
+    @Override
+    protected KeycloakId getId(UserModel userModel) {
+        return new KeycloakId(userModel.getId());
+    }
+
+    @Override
+    protected User scimRequestBodyForCreate(UserModel roleMapperModel) {
+        String firstAndLastName = String.format("%s %s", StringUtils.defaultString(roleMapperModel.getFirstName()),
+                StringUtils.defaultString(roleMapperModel.getLastName())).trim();
+        String displayName = Objects.toString(firstAndLastName, roleMapperModel.getUsername());
+        Stream<RoleModel> groupRoleModels = roleMapperModel.getGroupsStream().flatMap(RoleMapperModel::getRoleMappingsStream);
+        Stream<RoleModel> roleModels = roleMapperModel.getRoleMappingsStream();
+        Stream<RoleModel> allRoleModels = Stream.concat(groupRoleModels, roleModels);
+        List<PersonRole> roles = allRoleModels.filter(r -> BooleanUtils.TRUE.equals(r.getFirstAttribute("scim")))
+                .map(RoleModel::getName).map(roleName -> {
+                    PersonRole personRole = new PersonRole();
+                    personRole.setValue(roleName);
+                    return personRole;
+                }).toList();
+        User user = new User();
```

Comment:
For users, we could leverage user profile to configure the mapping between attributes in the `UserModel` to SCIM attributes.

I'm also wondering what are your thoughts about supporting SCIM schemas (e.g.: users) other than the core schema. Does it make sense as a follow-up?